### DISCUSSION TITLE
SNS

### DISCUSSION THREAD

**Discussion by Luzhiled** [luzhiled1333/comp-library#115] [General]
SNS ということです

なんでも書いてください

**Comment by ei1333** [comment]
やっほー

**Comment by Luzhiled** [comment]
R2 や R3 の整備を真面目にやっていきたいところだなという気持ちに

**Reply by ei1333** [reply]
まかせた

**Reply by Luzhiled** [reply]
頼んだ

**Reply by ei1333** [reply]
いや

**Comment by Luzhiled** [comment]
グラフが密かどうかとかあれやこれやって何か基準があったりするのかな / 判定はできるのかな

**Reply by Luzhiled** [reply]
噂によると Edge が $o(n^2)$ 本であるようなグラフクラスのことを疎グラフと呼ぶらしい　しらんけど

**Reply by ei1333** [reply]
実用的にはそのグラフで両方試して早い方を使いたいが、両方試すと

**Comment by Luzhiled** [comment]
verification-helper、100年くらいもすればなにか別のものに置き換わっていると思うんだけどね

**Reply by Luzhiled** [reply]
機能が十分に分割されていないのがまず悪さをしていそうというのはある
ドキュメントの生成用コマンドとして `oj-verify docs` が存在してはいるが、 `oj-verify run` のあとに実行しないと ❌ まみれになってしまうから実質 `oj-verify all` みたいなものなんだよな

**Reply by Luzhiled** [reply]
verify は十分なマシンスペックを仮定すれば並列実行できそうなもんだよね

**Reply by Luzhiled** [reply]
tag がつけられるような機能とかもあるとほんとは嬉しいですよね

**Reply by Luzhiled** [reply]
ライブラリをどう分類するべきか問題みたいなの、どうせ多対多なんだから多対多用の機能が実装されているほうがしっくりきます　僕は

**Reply by ei1333** [reply]
つまり verification-helper 2がほしいということだな

**Reply by Luzhiled** [reply]
そうだよ

**Reply by Luzhiled** [reply]
オタク検討会の結果、bazel とかで verify そのものは問題なく実行できそうということはわかってきた

**Reply by Luzhiled** [reply]
ただドキュメントの生成が結構面倒で、結局依存関係の解析を頑張る必要があり、終わりです

**Reply by Luzhiled** [reply]
verification-helper でも dependency graph (?) として計算していそうですね　このパートだけやってくれるものとかありそうなもんだけど

**Reply by ei1333** [reply]
ドキュメントは、いいか

**Reply by ei1333** [reply]
doxygenはどうだろう

**Reply by Luzhiled** [reply]
テストの実行時間の最大やメモリ使用量の最大を docs に自動で生成してほしい

timestamps.*.json がそれらの情報をもたないと `oj-verify run` で差分が発生してしまってつらいことになる

**Reply by Luzhiled** [reply]
https://github.com/luzhiled1333/comp-library/issues/79#issuecomment-1412319548

**Comment by ei1333** [comment]
紅茶の消費ペースをわかってないんだよね
アレクサにそろそろじゃないかと言われたんですが

**Reply by Luzhiled** [reply]
まあ実際そろそろではある

**Comment by Luzhiled** [comment]
noshi91 のライブラリの CI が壊れており、Twitter を見てみると困っていそうではあるが、僕が Twitter をやめているためリプが送れない

**Reply by Luzhiled** [reply]
ぱっと見 `GH_PAT` のアクセストークンの期限が切れているだけだと思うので PR とかで解決できる問題でもなさそうだな
誰か https://online-judge-tools.github.io/verification-helper/installer.ja.html を送ってあげてくれ

**Reply by ei1333** [reply]
わかった

**Comment by Luzhiled** [comment]
clang-format、せっかく `.clang-format` を親方向に再帰的に探して使ってくれるのに `-i $(find src/ test/ ...)` とかやっているせいでリポジトリの root ディレクトリでしかできないのなんとかしてえな

**Reply by Luzhiled** [reply]
```sh
(cd $(git rev-parse --show-toplevel) && clang-format -i $(find src/ test/ unit-test/ -name "*.hpp" -or -name "*.cpp"))
```
とかで動くか？

**Reply by Luzhiled** [reply]
これができるなら `oj-verify run` とかもそうしたほうがいいな

**Reply by Luzhiled** [reply]
submodule とかがあると破壊されるかもしれないが、まあ現状そんなものはないので OK だと信じよう

**Comment by Luzhiled** [comment]
PR template を試すには PR を送るしかないらしい

**Reply by ei1333** [reply]
PRを送ろう

**Reply by Luzhiled** [reply]
https://github.com/luzhiled1333/comp-library/issues?q=is%3Aissue+is%3Aopen+label%3A%22high+priority%22

**Comment by ei1333** [comment]
Luzhiled judge helperを作ろうという話

**Reply by ei1333** [reply]
https://github.com/competitive-verifier/competitive-verifier が気になっています

**Reply by kzrnm** [reply]
competitive-verifier/competitive-verifier の作者です。
検索でヒットしたので宣伝にきました。

## 機能

基本的には verification-helper の移植です。

verification-helper との大きな違いとしては「ソースコード解析」「verify」「ドキュメント生成」が独立した機能になったことです。JSONで結果をやり取りするようにしています。これによって、C# などのソースコード解析が重い言語では別ライブラリで解析を行えるようになったり、GitHub Actions での並列実行ができるようになったりしました。

ちなみに、昨日のアップデートでテストケースごとの実行結果も表示するようになりました。[example](https://competitive-verifier.github.io/competitive-verifier/examples/cpp/union_find.yosupo.test.cpp)

## 移行

competitive-verifier をローカルにインストールした後に、リポジトリ内で

```
competitive-verifier migrate
```

を実行すると verification-helper から移行できるようにしてるので移行は容易かと思います。

お二人のライブラリで試してみた結果です。

## ei1333/library

https://competitive-verifier.github.io/zzz_ei1333_library/

個別の変更点として、dmoj は実行できないので IGNORE を追加しています。
https://competitive-verifier.github.io/zzz_ei1333_library/test/verify/dmoj-ds5.test.cpp

## luzhiled1333/comp-library 

https://competitive-verifier.github.io/zzz_luzhiled1333_comp-library/

DROPBOX_TOKEN を設定してないので AtCoder に失敗してますが、ほかは実行できています。

**Comment by ei1333** [comment]
C++20にしようぜという話

**Reply by ei1333** [reply]
今はもう2023年なんですよね

**Comment by Luzhiled** [comment]
やっぱりドキュメントを書くパートとかちゃんとチェックするパートとかが露骨に面倒で、`src/zzz-gomi/` とかに verify もしなければドキュメントも書かれていないコード片を雑に置いていったほうがいいのではという気もしてくる

**Reply by Luzhiled** [reply]
いや、別に適切なディレクトリに配置していってもいいんですけどね

**Reply by Luzhiled** [reply]
PR の消化がネックで、 Projects で管理していきますか

**Reply by Luzhiled** [reply]
https://github.com/orgs/luzhiled1333/projects/3/views/1

**Comment by ei1333** [comment]
Pokémon Sleepで フレンドになりましょう！
わたしのリサーチャーコードはこちら
7582-8824-7053

**Reply by Luzhiled** [reply]
![Screenshot_20230815-191159](https://github.com/luzhiled1333/comp-library/assets/20723908/c2fb4f8f-51fc-4933-aed7-447f133b7b9d)

**Reply by ei1333** [reply]
お腹が空きませんか? Uber Eats で初回のご注文額が ￥1,900 以上の場合に ￥1,000 の割引を受けることができます。規約が適用されます。この紹介コードを会計時にお使いください: eats-8nd9eg7b8u   http://ubr.to/EatsGiveGet

**Reply by Luzhiled** [reply]
ban チャンス
